x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyExecutor.java

```diff
@@ -66,51 +75,103 @@ private void releasePolicy(String policyName) {
+    private EnrichPolicy getPolicy(ExecuteEnrichPolicyAction.Request request) {
         // Look up policy in policy store and execute it
-        EnrichPolicy policy = EnrichStore.getPolicy(policyId, clusterService.state());
+        EnrichPolicy policy = EnrichStore.getPolicy(request.getName(), clusterService.state());
         if (policy == null) {
-            throw new IllegalArgumentException("Policy execution failed. Could not locate policy with id [" + policyId + "]");
-        } else {
-            runPolicy(policyId, policy, listener);
+            throw new IllegalArgumentException("Policy execution failed. Could not locate policy with id [" + request.getName() + "]");
         }
+        return policy;
     }
 
-    public void runPolicy(String policyName, EnrichPolicy policy, ActionListener<PolicyExecutionResult> listener) {
-        tryLockingPolicy(policyName);
+    public void runPolicy(ExecuteEnrichPolicyAction.Request request, ActionListener<EnrichPolicyExecutionTask.Status> listener) {
+        runPolicy(request, getPolicy(request), listener);
+    }
+
+    public void runPolicy(ExecuteEnrichPolicyAction.Request request, TaskListener<EnrichPolicyExecutionTask.Status> listener) {
+        runPolicy(request, getPolicy(request), listener);
+    }
+
+    public Task runPolicy(ExecuteEnrichPolicyAction.Request request, EnrichPolicy policy,
+                          ActionListener<EnrichPolicyExecutionTask.Status> listener) {
+        return runPolicy(request, policy, (t, r) -> listener.onResponse(r), (t, e) -> listener.onFailure(e));
+    }
+
+    public Task runPolicy(ExecuteEnrichPolicyAction.Request request, EnrichPolicy policy,
+                          TaskListener<EnrichPolicyExecutionTask.Status> listener) {
+        return runPolicy(request, policy, listener::onResponse, listener::onFailure);
+    }
+
+    private Task runPolicy(ExecuteEnrichPolicyAction.Request request, EnrichPolicy policy,
+                           BiConsumer<Task, EnrichPolicyExecutionTask.Status> onResponse, BiConsumer<Task, Exception> onFailure) {
+        tryLockingPolicy(request.getName());
         try {
-            Runnable runnable = createPolicyRunner(policyName, policy, new PolicyUnlockingListener(policyName, listener));
-            threadPool.executor(ThreadPool.Names.GENERIC).execute(runnable);
+            return runPolicyTask(request, policy, onResponse, onFailure);
         } catch (Exception e) {
             // Be sure to unlock if submission failed.
-            releasePolicy(policyName);
+            releasePolicy(request.getName());
+            throw e;
+        }
+    }
+
+    private Task runPolicyTask(ExecuteEnrichPolicyAction.Request request, EnrichPolicy policy,
+                               BiConsumer<Task, EnrichPolicyExecutionTask.Status> onResponse, BiConsumer<Task, Exception> onFailure) {
+        Task asyncTask = taskManager.register("enrich", "", request);
```

Comment:
Technically we reuse the task that is being created in `TransportExecuteEnrichPolicyAction` and pass it down here,
but since we want to support wait for completion too then the logic here makes sense.

Reply:
What I think we can do here in order to make sure only a single `EnrichPolicyExecutionTask` is made is do the following:

```
TaskAwareRequest taskAwareRequest = new TaskAwareRequest() {
            @Override
            public void setParentTask(TaskId taskId) {
                request.setParentTask(taskId);
            }

            @Override
            public TaskId getParentTask() {
                return request.getParentTask();
            }

            @Override
            public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
                return new EnrichPolicyExecutionTask(id, type, action, getDescription(), parentTaskId, headers);
            }
        };
        Task asyncTask = taskManager.register("enrich", "policy_execution", taskAwareRequest);
```
 
and then remove the `createTask(...)` method in `ExecuteEnrichPolicyAction.Request` class.